### PR TITLE
Improve caching of anonymous types

### LINKED ISSUE #16511: Compiler hangs with 2.5.0-dev.20170613
It was working until 2.4.0.

https://travis-ci.org/falsandtru/spica/jobs/242650734

**TypeScript Version:**  2.5.0-dev.20170613

**Code**

https://github.com/falsandtru/spica

**Expected behavior:**

pass

**Actual behavior:**

```
$ gulp test
[09:29:50] Using gulpfile ...\spica\gulpfile.js
[09:29:50] Starting 'clean'...
[09:29:50] Finished 'clean' after 10 ms
[09:29:50] Starting 'test'...
[09:29:50] Starting 'ts:test'...

<--- Last few GCs --->

[8108:000001D865C4C640]   191743 ms: Mark-sweep 1401.1 (1504.4) -> 1401.1 (1504.4) MB, 5382.0 / 0.0 ms  allocation failure GC in old space requested
[8108:000001D865C4C640]   196651 ms: Mark-sweep 1401.1 (1504.4) -> 1401.1 (1473.4) MB, 4907.0 / 0.0 ms  last resort
[8108:000001D865C4C640]   201277 ms: Mark-sweep 1401.1 (1473.4) -> 1401.0 (1473.4) MB, 4624.0 / 0.0 ms  last resort


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 000002707D029891 <JS Object>
    1: set [native collection.js:~247] [pc=00000272AD3ED052](this=000000DA342CCA89 <a Map with map 00000362EC684AE9>,p=0000038C322A0099 <String[9]: 1653672:2>,x=0000038C322A00C1 <a TypeObject with map 00000362EC6E9249>)
    2: getTypeWithThisArgument [...\spica\node_modules\typescript\lib\typescript.js:~31548] [pc=00000272AD3BC751](this=000001933E08A311 <JS Global Ob...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

### LINKED ISSUE #17036: TSC crashes on <30 line project
Repro case: https://github.com/bcherny/tsoption/blob/7ae2d5d/index.ts ([tree](https://github.com/bcherny/tsoption/tree/7ae2d5d))

Repro steps:

1. Git pull tsoption@7ae2d5d
2. Run `npm install`
3. Run `./node_modules/.bin/tsc`

Node version: 6.9.5
TypeScript version: 2.4.1

Error output:

```sh
$ ./node_modules/.bin/tsc

<--- Last few GCs --->

   31467 ms: Mark-sweep 1339.6 (1437.6) -> 1339.6 (1437.6) MB, 1460.1 / 0.0 ms [allocation failure] [GC in old space requested].
   32931 ms: Mark-sweep 1339.6 (1437.6) -> 1339.6 (1437.6) MB, 1463.8 / 0.0 ms [allocation failure] [GC in old space requested].
   34405 ms: Mark-sweep 1339.6 (1437.6) -> 1340.7 (1421.6) MB, 1473.7 / 0.0 ms [last resort gc].
   35885 ms: Mark-sweep 1340.7 (1421.6) -> 1341.8 (1421.6) MB, 1480.1 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x11cf6b9cfb51 <JS Object>
    1: getReturnTypeOfSignature [/Users/bcherny/Sites/tsoption/node_modules/typescript/lib/tsc.js:~27029] [pc=0x11e2ac7e930b] (this=0x11cf6b9e6f19 <JS Global Object>,signature=0x35a7fc95c541 <a Signature with map 0x210c46c5c571>)
    2: getReturnTypeOfSignature [/Users/bcherny/Sites/tsoption/node_modules/typescript/lib/tsc.js:~27029] [pc=0x11e2ac7e84ad] (this=0x11cf6b9e6f19 <JS Global Object>,s...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/Users/bcherny/.nvm/versions/node/v6.9.5/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/Users/bcherny/.nvm/versions/node/v6.9.5/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/Users/bcherny/.nvm/versions/node/v6.9.5/bin/node]
 4: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/Users/bcherny/.nvm/versions/node/v6.9.5/bin/node]
 5: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/bcherny/.nvm/versions/node/v6.9.5/bin/node]
 6: 0x11e2ac2079a7
Abort trap: 6
```

### LINKED ISSUE #17968: tsc / language server hangs with complex type aliases
**TypeScript Version:**  2.4.2

**Code**

Originally, trying to abuse the type system to create guaranteed-balanced AVL trees. Found out that language server used with vscode would just hang or crash, so I reduced it slightly, and tried running it through `tsc`:

```ts
type NonNeg = {sign: "0"} | {sign: "+", pred: NonNeg}
type NonPos = {sign: "0"} | {sign: "-", succ: NonPos}
type Int = {sign: "0"} | {sign: "+", pred: NonNeg} | {sign: "-", succ: NonPos}

type Succ<N extends Int> = {"0": {sign:"+", pred:{sign:"0"}}, "+" :{sign:"+", pred:N}, "-": N["succ"]}[N["sign"]];
type Pred<N extends Int> = {"0": {sign:"-", succ:{sign:"0"}}, "-" :{sign:"-", succ:N}, "+": N["pred"]}[N["sign"]];

type Zero = {sign: "0"};

type Empty<N extends Int> = {z: {}, s: never, n: never}[N["value"]] & {type: "empty", level: Zero};
type Even<N extends Int> = {type: "even", left: AVL<Pred<N>>, right: AVL<Pred<N>>, value: number, level: N};
type LeftHeavy<N extends Int> = {type: "left-heavy", left: AVL<Pred<N>>, right: AVL<Pred<Pred<N>>>, value: number, level: N};
type RightHeavy<N extends Int> = {type: "right-heavy", left: AVL<Pred<Pred<N>>>, right: AVL<Pred<N>>, value: number, level: N};

type AVL<N extends Int> = Even<N> | LeftHeavy<N> | RightHeavy<N> | Empty<N>

// note: this function is mostly nonsensical, but it appears to be the cause of the problem
function find<N extends Int>(x: number, tree: AVL<N>): void {
    if (tree.type == "empty") {
        return;
    }
    find(x, tree.left);
}
```

**Expected behavior:**

Accept, probably (assuming I made no mistakes), but a compile error would be fine too.

**Actual behavior:**

`tsc` hangs for a couple minutes minutes, then prints a stack trace indicating that v8 ran out of memory:

```
<--- Last few GCs --->

[71249:0x103801600]    94152 ms: Mark-sweep 1401.9 (1582.3) -> 1401.9 (1554.3) MB, 1944.7 / 0.0 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 1945 ms) last resort 
[71249:0x103801600]    96016 ms: Mark-sweep 1401.9 (1554.3) -> 1401.9 (1554.3) MB, 1863.1 / 0.0 ms  last resort 


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x36703af9bbd9 <JS Object>
    1: set [native collection.js:~247] [pc=0x86dbef1a89a](this=0x3db4171451b1 <a Map with map 0x2fd6c9514319>,p=0x227e7acccf51 <String[15]: 1771856,1771870>,x=1)
    2: /* anonymous */ [/usr/local/lib/node_modules/typescript/lib/tsc.js:978] [pc=0x86dbef0f2c9](this=0x9de99b0a1e1 <JS Global Object>,value=1,key=0x227e7acccf51 <String[15]: 1771856,1771870>)
    3: arguments adaptor frame: 3->2
    ...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 5: v8::internal::Factory::NewFixedArray(int, v8::internal::PretenureFlag) [/usr/local/bin/node]
 6: v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, v8::internal::JSMapIterator, 2>::Allocate(v8::internal::Isolate*, int, v8::internal::PretenureFlag) [/usr/local/bin/node]
 7: v8::internal::OrderedHashTable<v8::internal::OrderedHashMap, v8::internal::JSMapIterator, 2>::Rehash(v8::internal::Handle<v8::internal::OrderedHashMap>, int) [/usr/local/bin/node]
 8: v8::internal::Runtime_MapGrow(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
 9: 0x86dbea840bd
10: 0x86dbef1a89a
11: 0x86dbef0f2c9
12: 0x86dbea86bbb
13: 0x86dbef09cab
Abort trap: 6
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
We have recently seen a number of out-of-memory issues that are caused by runaway anonymous object type instantiations and comparisons. This PR introduces a new and much improved caching scheme for such types. The key idea is to track which type variables are in scope at the source origin of each anonymous object type and cache instantiations of the type based on the actual type arguments for those type variables. This means anonymous object type instantiation caching becomes as efficient as that for classes and interfaces, and out-of-memory issues that were previously fixed by switching from type aliases to interfaces should now be gone.

The RWC tests all pass. Only changes are some missing elaborations in error messages that result from improved sharing of identical types.

Fixes #16511.
Fixes #17036.
Fixes #17968.

### CODE REVIEW

**Comment by sandersn** [comment]
We should get some memory numbers for large or medium projects that don't benefit from this (if there are there are projects like this) to make sure there's not too much overhead.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8174,13 +8178,53 @@ namespace ts {
...
-            return result;
+        function getAnonymousTypeInstantiation(type: AnonymousType, mapper: TypeMapper) {
+            const target = type.objectFlags & ObjectFlags.Instantiated ? type.target : type;
+            const symbol = target.symbol;
+            const links = getSymbolLinks(symbol);
+            let typeParameters = links.typeParameters;
+            if (!typeParameters) {
+                // The first time an anonymous type is instantiated we compute and store a list of the type
+                // parameters that are in scope (and therefore potentially referenced). For type literals that
+                // aren't the right hand side of a generic type alias declaration we optimize by reducing the
```
Comment (sandersn) [inline src/compiler/checker.ts]: should be "the rhs of a non-generic type alias declaration", right?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -8174,13 +8178,53 @@ namespace ts {
...
+            return type;
+        }
+
+        function isTypeParameterReferencedWithin(tp: TypeParameter, node: Node) {
+            return tp.isThisType ? forEachChild(node, checkThis) : forEachChild(node, checkIdentifier);
+            function checkThis(node: Node): boolean {
+                return node.kind === SyntaxKind.ThisType || forEachChild(node, checkThis);
+            }
+            function checkIdentifier(node: Node): boolean {
+                return node.kind === SyntaxKind.Identifier && isPartOfTypeNode(node) && getTypeFromTypeNode(<TypeNode>node) === tp || forEachChild(node, checkIdentifier);
```
Comment (sandersn) [inline src/compiler/checker.ts]: This call to getTypeFromTypeNode is likely to 'pull' a lot of types that might otherwise never be instantiated themselves. It's probably nothing because, during instantiation, we'll surely be instantiating nearby types anyway, but worth thinking about nonetheless.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -8174,13 +8178,53 @@ namespace ts {
...
+                // set of type parameters to those that are actually referenced somewhere in the literal.
+                const declaration = symbol.declarations[0];
+                const outerTypeParameters = getOuterTypeParameters(declaration, /*includeThisTypes*/ true) || emptyArray;
+                typeParameters = symbol.flags & SymbolFlags.TypeLiteral && !target.aliasTypeArguments ?
+                    filter(outerTypeParameters, tp => isTypeParameterReferencedWithin(tp, declaration)) :
+                    outerTypeParameters;
+                links.typeParameters = typeParameters;
+                if (typeParameters.length) {
+                    links.instantiations = createMap<Type>();
+                    links.instantiations.set(getTypeListId(typeParameters), target);
```
Comment (sandersn) [inline src/compiler/checker.ts]: I don't understand how caching the uninstantiated type helps.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -4794,30 +4794,47 @@ namespace ts {
...
+                    case SyntaxKind.ConstructorType:
+                    case SyntaxKind.JSDocFunctionType:
+                    case SyntaxKind.FunctionDeclaration:
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.FunctionExpression:
+                    case SyntaxKind.ArrowFunction:
+                    case SyntaxKind.TypeAliasDeclaration:
+                    case SyntaxKind.JSDocTemplateTag:
+                    case SyntaxKind.MappedType:
+                        const outerTypeParameters = getOuterTypeParameters(node, includeThisTypes);
```
Comment (sandersn) [inline src/compiler/checker.ts]: Why is this line needed? Aren't further iterations of the while loop equivalent to calling getOuterTypeParameters recursively?

Comment (sandersn) [reply src/compiler/checker.ts]: Oh, because we always return whenever any of these cases is true. Why not append to a local list, break, and keep looping instead of recurring and returning? Seems like it would be more efficient (as long as the final result isn't order-dependent, but that could be fixed by swapping the order in calls to `append`).

**Comment by sandersn** [comment]
Also fixes the crash found by @mulyoved when using glamorous, which used the complicated `Omit` mapped type.

File: src/compiler/checker.ts (hunk 4 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: Our use of the `getOuterTypeParameters` function in code for classes and interfaces requires the list to be in outside-in order. Hence the recursion. We could insert new elements at the beginning of the list or reverse the list at the end instead, but neither are better solutions.

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: No, we optimize for type literals, except when they are the RHS of a generic type alias. For those we want to preserve the distinct instantiations even when they don't reference the type parameters (for fidelity when displaying types).

File: src/compiler/checker.ts (hunk 3 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: The uninstantiated type is the same as an instantiation where each type parameter has itself as its type argument. No point in recomputing that instantiation, we already have it.

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: No, it is always fine to call `getTypeFromTypeNode`. It will be called by the checker anyways when as we type check a program.

**Comment by ahejlsberg** [comment]
@sandersn Regarding memory usage, it varies from no change to massive savings, but I haven't seen meaningfully increased consumption anywhere. There is no significant change for the Monaco and TFS code bases, for example.
